_posts/2020-06-14-dynamic-partition-pruning.md

```diff
@@ -0,0 +1,159 @@
+---
+layout: post
+title:  "Dynamic partition pruning"
+author: Raunaq Morarka, Qubole and Karol Sobczak, Starburst Data
+excerpt_separator: <!--more-->
+---
+
+Improving dynamic filtering to speed up inner-joins on partitioned tables.
+
+<!--more-->
+
+# Introduction
+
+Consider the following query which captures a common pattern of a fact table `store_sales` partitioned by the column 
+`ss_sold_date_sk` joined with a filtered dimension table `date_dim`.
+```
+SELECT COUNT(*) FROM 
+store_sales JOIN date_dim ON store_sales.ss_sold_date_sk = date_dim.d_date_sk
+WHERE d_following_holiday='Y' AND d_year = 2000;
+```
+Without dynamic filtering, Presto will push predicates for the dimension table to the table scan on `date_dim` but 
+it will scan all the data in the fact table since there are no filters on `store_sales` in the query.
+The join operator will end up throwing away most of the probe-side rows as the join criteria is highly selective. 
+The current implementation of [dynamic filtering](https://prestosql.io/blog/2019/06/30/dynamic-filtering.html) improves on
+this by applying dynamic filtering to broadcast joins on tables stored in ORC or Parquet format. 
+But it does not take advantage of the layout of partitioned Hive tables.
+
+With dynamic partition pruning which extends the current implementation of dynamic filtering, every worker node collects 
+values eligible for the join from `date_dim.d_date_sk` column and passes it to the coordinator where the processing 
+of the partitions of `store_sales` which don't meet the join criteria can be skipped to greatly reduce the amount of data
+scanned from `store_sales`.
+
+# Design considerations
+
+We needed to ensure that the coordinator is not overloaded by too many updates for dynamic filters sent from worker nodes.
+This was achieved by packing dynamic filters into Presto's existing framework for sending status updates from worker to coordinator.
+
+[`DynamicFilterService`](https://github.com/prestosql/presto/blob/master/presto-main/src/main/java/io/prestosql/server/DynamicFilterService.java) 
+was added on the coordinator node to perform dynamic filter collection asynchronously so that
+queries registered with this service can request dynamic filters while scheduling splits without blocking any operations.
+This service is also responsible for ensuring that all the build-side tasks of a join stage have completed execution before 
+constructing dynamic filters for use in the scheduling of probe-side scans from the coordinator.
+
+# Implementation
+
+For identifying opportunities for dynamic filtering in the logical plan, we rely on the implementation added in
+[#91](https://github.com/prestosql/presto/pull/91). Dynamic filters are modeled as `FunctionCall` expressions which 
```

Comment:
I wonder if we shouldn't just link to https://prestosql.io/blog/2019/06/30/dynamic-filtering.html here for more details about predicate push down and dynamic filters

Reply:
That blog post was written when the initial PR was completely independent of the planner (it says `there is no dependencies on the planner` in the blog). Later on as part of reviews that changed evolved to depend on DF planner. So I thought it would be interesting to provide an overview of the changes done in planner here.

Reply:
>  So I thought it would be interesting to provide an overview of the changes done in planner here.

Sounds good to me